### ISSUE TITLE
[Tracking] Add performance CI tests

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#11711]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Open a pull request in NewDot.

## Expected Result:
A suite of performance tests should run, so that we know _before a PR is merged_, if it creates a statistically significant performance regression.

## Actual Result:
We don't have performance CI tests

## Workaround:
_Just don't break stuff™_

## Platform:
n/a

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by roryabraham** [comment]
cc @hannojg Can you please comment on this so I can assign it to you

**Comment by hannojg** [comment]
Yes 👋

**Comment by AndrewGable** [comment]
PR was merged, however, we had a lot of testing to confirm a few questions:
1. Does the PR work?
2. Can we get the "Large Runners" working?
3. Do the "Large Runners" speed things up?
4. Do the "Large Runners" have KVM support to allow hardware acceleration?

**Comment by AndrewGable** [comment]
After testing via this PR:
1. PR works, but required some small edits I made here: https://github.com/Expensify/App/pull/11936
2. Large runners work!
3. Large runners seem to speed up most things, but not the tests.
4. Large runners do not seem to have kvm support out of the box, but @hannojg might be able to enable them. We will discuss further.

**Comment by AndrewGable** [comment]
**Problem**: We are trying to configure cloud machines to run an emulator very quickly to run performance tests. However:
> Emulators can be fast, but require hardware acceleration (HAXM on Mac & Windows, KVM on Linux) from the host. This presents a challenge on CI especially when running hardware accelerated emulators within a docker container, because Nested Virtualization must be supported by the host VM which isn't the case for most cloud-based CI providers due to infrastructural limits. ([source](https://github.com/ReactiveCircus/android-emulator-runner/pull/279/files))
 
**Solution**: Skip running the tests on emulators on CI machines and run the tests on physical devices via AWS Device Farm. 

I'm still validating this idea is possible, but I've confirmed we can at least start the tests via AWS Device Farm. I think the real question will be if they are fast enough to get reliable results.

<img width="1519" alt="Screen Shot 2022-10-19 at 7 27 29 PM" src="https://user-images.githubusercontent.com/2838819/196837130-f9459ce7-318f-4c20-98ab-883c9b75d96d.png">

Once we prove the POC, we can hook it up to [`realm/aws-devicefarm`](https://github.com/realm/aws-devicefarm) which provides an easy to use GitHub Actions workflow.

**Comment by AndrewGable** [comment]
Discussing in Slack with @hannojg - But I was unable to get the custom node http server working on the aws device farm machines. I think looking to see if Appium can support our tests might be the best next move since they are supported out of the box by AWS device farm.

**Comment by AndrewGable** [comment]
I've reached out to AWS Device Farm support via a support case, I have chatted with their support on the phone and provided code details. They are circling back internally and I will let you know when I hear back from them.

**Comment by hannojg** [comment]
My guess after trying hard yesterday to make our simple HTTP server impl working is that it would work with a "private device" for which we can setup a VPC https://docs.aws.amazon.com/devicefarm/latest/developerguide/amazon-vpc-endpoints.html.

However, that might not be feasible as setup, I am currently migrating to appium.

**Comment by AndrewGable** [comment]
@hannojg and I are getting on a call with AWS device farm team today

**Comment by AndrewGable** [comment]
Update from our call with AWS Device farm:
1. They gave us a code snippet that enables us to run the Android tests as we'd like 🎉 I am working on getting the tests working as originally written now. If I cannot get them to work today, I will hand this off to @hannojg for tomorrow.
2. AWS Device farm let us know however that the way we are running these tests will not be able to be run on iOS devices. This is because iOS restricts the ability to manipulate the network on iOS devices starting on iOS 12 (very old OS version).
3. I would think we can start with Android devices as they tend to be slower, than if we want to build this out we can look at their recommended solution ([VPC](https://docs.aws.amazon.com/devicefarm/latest/developerguide/amazon-vpc-endpoints.html)).

**Comment by AndrewGable** [comment]
We had our first successful run via AWS device farm for Android e2e tests 🎉 Now we just need to improve and tweak things.

```
❇️  Performance comparison results:

➡️  Significant changes to duration

➡️  Meaningless changes to duration
 - TTI: 1560.9 ms → 1605.7 ms (+44.8 ms, +2.9%) 
 - runJsBundle: 232.0 ms → 241.4 ms (+9.4 ms, +4.0%) 
 - regularAppStart: 0.0 ms → 0.0 ms (+0.0 ms, +8.5%) 
 - nativeLaunch: 16.7 ms → 15.9 ms (-0.8 ms, -4.7%) 
```

**Comment by AndrewGable** [comment]
I've done 1 & 2 on this branch/fork: https://github.com/AndrewGable/App/pull/1

I think the last part of this will be figuring out how to configure AWS to report back the results. I believe this can be done using the "artifacts" as mentioned here: https://github.com/realm/aws-devicefarm#download-artifacts

**Comment by AndrewGable** [comment]
Here is an example of a passing test run: https://github.com/AndrewGable/App/actions/runs/3382629912/jobs/5617747052

It took 45 minutes. I think we can definitely figure out how to get this working more efficiently, but we can focus on that after step 3 & 4 are done.

**Comment by AndrewGable** [comment]
Spent most of the morning attempting different paths for artifacts, but wasn't able to get it to work. I reached out to our contact at AWS via email hoping that he can point us in the right direction.

**Comment by AndrewGable** [comment]
Artifacts are now working, just doing some final touches on how to comment the results back from AWS here: https://github.com/AndrewGable/App/pull/1

**Comment by AndrewGable** [comment]
Next steps list:

- [ ] Get commenting on PR working with output from AWS
- [ ] Build two APKs via GitHub Actions to compare (right now it's comparing the same build against itself)
- [ ] Upload both APKs via AWS GitHub Actions workflow
- [ ]  Speed up the tests without introducing false negatives
    - [ ] Re-enable large runners
    - [ ] Speed up the AWS tests

**Comment by AndrewGable** [comment]
First successful comment run here: https://github.com/AndrewGable/App/pull/1#issuecomment-1303012146 🎉

**Comment by hannojg** [comment]
Looks sexy!

<img width="458" alt="Screenshot 2022-11-04 at 07 43 02" src="https://user-images.githubusercontent.com/16821682/199908647-e2be5caf-db18-4e4d-a5ed-6b05c14661b9.png">

Thank you for finishing up that system Andrew, great work 💪

**Comment by AndrewGable** [comment]
Ran into all Android builds breaking today: https://expensify.slack.com/archives/C01GTK53T8Q/p1667584299633799

Wasn't able to fix this error so it halted development. I will try again next week to finish the next steps list.

**Comment by AndrewGable** [comment]
Back on this 🐴 - Here is a run from two different branches: https://github.com/AndrewGable/App/pull/1#issuecomment-1306338309

**Comment by AndrewGable** [comment]
Getting closer to having something that is able to be merged! I think a few more days of testing should get us there.

**Comment by JmillsExpensify** [comment]
Woo! Perfect timing on your update.

**Comment by AndrewGable** [comment]
Moved my PR over to Expensify/App instead of my own fork, need to get the secrets set up on this repo so I created a Ring0 issue linked above.

**Comment by AndrewGable** [comment]
Ok got tests running on this repo, but I am running into a funny error. I change the code locally but it seems to not "reflect" on the AWS side. I am thinking that we might need to clean up our uploads to AWS when a test is complete, as this error doesn't seem to happen when I manually delete the uploads. 

I will work on this tomorrow.

**Comment by AndrewGable** [comment]
Ok @hannojg we have a working PR! https://github.com/Expensify/App/pull/12320#issuecomment-1309004239

However, I am seeing some "false positives" reported. Do you have any ideas on how to solve? I am running on basically the same branch to compare against each other.

**Comment by AndrewGable** [comment]
Should we be throwing out any outliers? (I need to look at the code and see if it's already doing that) This specific one looks like it had a test take 5x too long and throw things off.

<img width="765" alt="Screenshot 2022-11-09 at 9 24 33 AM" src="https://user-images.githubusercontent.com/2838819/200885068-0d7b372b-f9d0-4720-921e-235446cad75d.png">

**Comment by AndrewGable** [comment]
Next steps:
- [ ] Figure out how to prevent false positives
- [ ] Increase speed of test
     - First APK takes ~17 minutes to build
     - Second APK takes ~7 minutes to build
     - Tests take ~15 minutes, only ~3 minutes of that time is the actual test running. I asked AWS for our options here.

**Comment by hannojg** [comment]
Outliers are already removed, yes. I think it will be most worth investigating why those long times are happening. Maybe it's waiting for an API call something. Will investigate as well ✌️

**Comment by AndrewGable** [comment]
I see that we are using a fixed number of outliers (8 total, 4 on each side), I am going to try looking at an alternative filter via IDR, e.g. https://stackoverflow.com/a/45804710/1858217

Agree we probably should understand why the outlier exists too.

**Comment by AndrewGable** [comment]
Both https://github.com/Expensify/App/pull/12320#issuecomment-1309250069 and https://github.com/Expensify/App/pull/12320#issuecomment-1309327702 are using the new IQR method and the results look to have less outliers. Thoughts @hannojg ?

**Comment by AndrewGable** [comment]
1. Still discussing with AWS contact about how to speed device farm up
2. Still waiting on internal configuration of larger computer to speed up APK build time

I'll circle back here when we have any new details.

**Comment by AndrewGable** [comment]
FYI I tested 30 vs 15 iterations:
* 30 iterations: total time ~44.5 minutes, test time ~14.5 minutes on test
* 15 iterations: total time ~48.5 minutes, test time ~10 minutes on test

**Comment by AndrewGable** [comment]
@rafecolton enabled the large runners again for us, seeing how fast the tests are now 🚀

**Comment by AndrewGable** [comment]
Large runner results:
1. First APK took ~9 minutes (compared to ~17 minutes to build)
2. Second APK took ~4 minutes (compared to ~7 minutes to build)
3. Tests take about ~14 minutes (compared to ~15 minutes)

Total time ~30 minutes vs ~45 minutes

**Comment by JmillsExpensify** [comment]
Nice improvement!

**Comment by AndrewGable** [comment]
I still worry this is too long to roll out in it's current iteration, our current tests run in under 5 minutes. I don't think we can expect to run these tests in that time, but I think we need to get them under a certain amount of time or increase their value.

**Comment by JmillsExpensify** [comment]
Totally. So what do you think that number is out of curiosity? 10? 15? Something else?

**Comment by AndrewGable** [comment]
If we are going to run it on every commit, I think ~10 minutes would be OK. I don't know if it's realistic we will ever be able to speed these tests up that much so I think we might have to do something "out of the box". 

One idea is we could run these tests on merge, then report the results back to the PR. If they introduce a performance regression then we could revert the PR, or something similar. Otherwise we could run these via a label, or only on large PRs, etc. I don't have an exact proposed solution yet, but curious for thoughts from @Expensify/mobile-deployers @marcaaron ?

**Comment by JmillsExpensify** [comment]
> One idea is we could run these tests on merge, then report the results back to the PR. If they introduce a performance regression then we could revert the PR, or something similar.

I would love to have the most coverage possible, so this is an ideal place to start. Given that time between merge and QA is most cases, the 10 min "delay" also feels OK. In any case, not the real decision maker in this department. 😄

**Comment by AndrewGable** [comment]
Maybe a more refined proposal would be:
1. Run any long running tests after merge in the "Process new code merged to main" workflow
2. When the perf tests finish, add a comment to the PR mentioning the results
3. If there is a regression found ("Significant Changes To Duration") then add the deploy blocker label to the PR with a comment like "There was a performance regression found on this PR, please investigate with priority"
4. This will be added to the deploy checklist and we will not deploy until it's resolved

**Comment by roryabraham** [comment]
Sounds good @AndrewGable, but can we please make sure that, unlike [these post-merge test and lint jobs](https://github.com/Expensify/App/blob/e74400679f70a2da46e2f07d5e0449e6b6775852/.github/workflows/preDeploy.yml#L8-L12) we already have, the long-running tests do not block the deploy until they finish? i.e: assume that PRs do not introduce a performance regression, deploy them to staging, and then only if they do not pass post a comment and mark them as a deploy blocker

**Comment by AndrewGable** [comment]
End of week update: We have a workable solution [here](https://github.com/Expensify/App/issues/11711#issuecomment-1312139072), but still talking with AWS to juice up the device farm side of things.

**Comment by AndrewGable** [comment]
I am attempting to split the e2e tests up from the rest of the source code to reduce the size of the ZIP we send to the Android phone via AWS. However, I have been blocked today by a [broken Android build](https://expensify.slack.com/archives/C01GTK53T8Q/p1668457456678299).

**Comment by AndrewGable** [comment]
I was able to squeeze some more performance improvements out of the test, total time is now ~23 minutes:
1. First APK took ~9 minutes (compared to ~17 minutes to build)
2. Second APK took ~4 minutes (compared to ~7 minutes to build)
3. Tests take about ~9 minutes (compared to ~15 minutes)

I will work on the final piece which is running this in the "Process new code merged to main" workflow.

**Comment by AndrewGable** [comment]
Running into an issue with linting as described here: https://github.com/SchemaStore/schemastore/issues/2579

**Comment by AndrewGable** [comment]
Trying to finish this right now, next steps are:
- [ ] Call from external workflow (`preDeploy.yml`)
- [ ] Leave the deploy blocker label when there is `Significant Changes To Duration`

**Comment by AndrewGable** [comment]
PR finally in review 😮‍💨

**Comment by JmillsExpensify** [comment]
Woo! P.S. I didn't cover this in the regular open source update today but was going to profile it in the margelo room tomorrow.

**Comment by AndrewGable** [comment]
Nice, confirmed these are now running on merge into `main`: https://github.com/Expensify/App/pull/13570#issuecomment-1349908084

**Comment by AndrewGable** [comment]
These are running, there are some further separate improvements to make (e.g. adding more tests and mocking API responses), but I am going to consider this one done for now.
